Multirobot SLAM ros2

acachathuranga opened the pull request:
<!-- Please fill out the following pull request template for non-trivial changes to help us process your PR faster and more efficiently.-->

---

## Basic Info

An extension of slam_toolbox for ROS2 multi-robot distributed SLAM. This work was done as part of a research at Singapore University of Technology and Design. 

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | [(76)](https://github.com/SteveMacenski/slam_toolbox/issues/76) |
| Primary OS tested on | (Ubuntu 24.04) |
| Tested on gazebo and indoor space at Singapore University of Technology and Design | Gazebo and Custom  made Physical Robot (Monsterborg platform with RPLidar A3) |

---

## Description of contribution in a few bullet points

* Extends slam_toolbox to perform multi-robot mapping and localization in ROS2 (Can also be used for multi-laser scenarios).
* multirobot_slam_toolbox_node implements the new functionality separately (Behaviors and functionalities of all other slam_toolbox nodes remain unchanged).
* Reduces network bandwidth usage by only sharing pose-graph scan data between each robot (Uses new message type LocalizedLaserScan). No sharing of any other topics or data between the robots is required.
* Constructs a merged pose-graph and performs loop closure distributively on each robot.
* Not the original use-case, but this can also be used in multi-laser robots, by running a slam_toolbox instance for each laser.

![multi-robot_slam](https://github.com/acachathuranga/slam_toolbox/blob/226bb1a872f905eb72ec6c69c5f8bb895941671c/images/multi-robot_mapping.gif?raw=true "Mapping Image")

## Description of documentation updates required from your changes

* Slam_toolbox instances needs to be run on each robot under unique namespaces.
* New topic '/localized_scan' to share pose-graph data between robots. This topic must not be in a namespace.
* Each slam_toolbox instance publishes transform of host robot and peers in the network. Peer transform frame names are prefixed with the namespace of the peer.
* Each laser is named by the host namespace.

![multi-robot_architecture](https://github.com/acachathuranga/slam_toolbox/blob/226bb1a872f905eb72ec6c69c5f8bb895941671c/images/multi-robot_architecture.png?raw=true "Mapping Architecture")

---

## Future work that may be required in bullet points
* Experiment on lossy WiFi networks. How would the nodes handle scan message losses between each others.

acachathuranga committed:
Merge upstream changes from slam_toolbox/ros2

acachathuranga committed:
Added namespace to laser name

acachathuranga committed:
laser orient calculation using tf multiple

acachathuranga committed:
moved laser pose reading to seperate function

acachathuranga committed:
moved laser pose reading to seperate function

acachathuranga committed:
Added multirobot_async node

acachathuranga committed:
Add localizedScan publishing

acachathuranga committed:
Add localizedScan subscription and helper methods

acachathuranga committed:
Reduce ceres solver thread count

acachathuranga committed:
external robots tf publishing

acachathuranga committed:
Changed to odometry pose sharing instead of corrected pose

acachathuranga committed:
Ceres solver thread count revert

acachathuranga committed:
added architecture image and mapping gif

Daniel-Meza commented:
@acachathuranga this looks great.

Could you go into more detail about how you set up the two robot scenario and run the multirobot_slam_toolbox_node?
Looking at your forked multirobot_ros2 repo I don't see launch files to use as reference. If you have the ones you used for the demo above that would also be very helpful!

FYI, I am trying to replicate your demo using two Turtlebot3 robot platforms in Gazebo. I have set up the namespaces according to your diagram but when launching everything I can't seem to get the correct /robot1/map and /robot2/map topics to publish and the /localized_scan topic shows no published messages.

Appreciate any pointers.

acachathuranga committed:
added multirobot launch file

acachathuranga committed:
Added turtlebot multirobot mapping instructions

acachathuranga commented:
@Daniel-Meza I have added the launch and parameter files for multi-robot mapping. The parameters file is identical to mapper_params_online_async.yaml. 

I have also updated the ReadMe on how to run multi-robot slam with turtlebot3 multirobot simulation (provided with navigation2 nav2_bringup package). 

In short, you can follow the installation instructions in [navigation2 Getting started](https://navigation.ros.org/getting_started/index.html) tutorial. Once you've installed navigation2, nav2_bringup and turtlebot3 packages, source the environment variables as described and run the following commands in separate terminals.

```shell
ros2 launch nav2_bringup multi_tb3_simulation_launch.py headless:=False
ros2 launch slam_toolbox online_async_multirobot_launch.py namespace:=robot1
ros2 launch slam_toolbox online_async_multirobot_launch.py namespace:=robot2
```
Then you can use teleop_twist_keyboard to drive the robots around using following commands. When driving the robots around, the merged map will be displayed on both rviz windows.

```shell
ros2 run teleop_twist_keyboard teleop_twist_keyboard __ns:=/robot1
ros2 run teleop_twist_keyboard teleop_twist_keyboard __ns:=/robot2
```

acachathuranga committed:
Added turtlebot multirobot mapping instructions

Daniel-Meza commented:
@acachathuranga Thank you, I was able to replicate the demo and also get it working with a 3rd robot.

Going a bit further, do you have some insights on the following:

**Maximum number of robots.** Since every robot/sensor runs its own slam_toolbox node, do you foresee any limitation on the number of robots that can be used simultaneously? Assuming of course there is enough computing power and bandwidth to handle all of the data.

**Running on real robots.** In Gazebo simulation, I believe the "map" and "odom" frame is always placed at a global (0, 0, 0) regardless of where the robots are spawned. This means that every slam_toolbox node has their "map" located based on the same initial point. This is not the case when booting up real robots since the map frame is placed on the current physical location of the robot itself. So, would it be necessary to provide an additional transform to ensure that the "map" frame of real robots align on top of each other? Or is this transformation handled automatically by the multirobot nodes?

acachathuranga commented:
@Daniel-Meza nice to hear you got it working!

**Maximum number of robots:** Given there is enough computing power and bandwidth, I do not see any limitation on the number of robots (Unless there are underlying SLAM limitations within slam_toolbox on number of graph nodes etc). But nothing specific to 'multi-robot slam'. However, when the robot count increases in real world scenarios (at least during our testing), you're likely to run into bandwidth problems before anything else :)

**Running on real robots:** This is explained in the PR notes as well as the updated ReadMe. To summarize; gazebo simulation odometry is indeed in a global frame. So everything will automatically work. When you run on actual robots, you need to start the robots very close to each other, and also in the same orientation. Then the scan matcher will handle the initial offsets. In our testing, what we did was, 1) turn on one robot at a marked location, 2) move the robot away from the location 3) place another robot at the marked location and repeat. Taking initial poses from input args /parameters will be added in near future.

i1Cps commented:
Good Morning, Are you able to incoperate the recently added /reset feature with the multi-robot solution?

acachathuranga commented:
@SteveMacenski any thoughts or comments on this? I noticed many people find this multi-robot implementation useful and there have also been some recent duplicate efforts (PR  #727) on this multi-robot slam topic. 

This work has been tested extensively with real robot fleets (2 robot, 4 robot, 8 robot configurations) and has provided very good results. The use of '/localized_scan' topic instead of sharing all laser scans among all robots provides real-life benefits in terms of network bandwidth. We also did extensive tests with different types of network (Infrastructure based, mesh .etc) and found out sharing all laser messages is indeed not practical for real situations. We already have done further extensions on this work, in terms of bandwidth optimizations for large robot fleets. I will push these updates once our work is published.

MikeDegany commented:
Is it possible to share robots initial poses (relative poses) before multi-robot mapping starts, so that the robots do not need to start from the same location?

SteveMacenski commented:
Sorry for the delays here - this is something I'm going to prioritize responses and a resolution to now. Expect more responses from me now :-) 

There's another PR open by @Daniel-Meza which adds this into `humble` in a different way (ie the nodes themselves have multirobot support added instead of adding in a separate node) in https://github.com/UTNuclearRoboticsPublic/slam_toolbox/tree/humble_multirobot_v2.

I would like to setup a call with Daniel, you @acachathuranga and myself to chat about steps forward here so we can get some solution into `ros2` and possibly backported into Humble / Jazzy. We have three suggested ways (this #727 and the branch I liked above) and I'd like for us to pick the way we want to go; each test it for our needs; review it so that we're all happy and merge for a release. I think its easier to do that with a synchronous meeting so there's no delay. 

I possibly see a solution being 2 fold: your new node here and Daniel's work to embed the optionality into the single node for less networking complex situations / simulation. Or if we all agree one way in particular is the right way to go for all cases, then we can go that route and work together to make that happen.

Daniel-Meza commented:
Sounds good, @SteveMacenski.
I can usually find time mid-day during the week. If you could suggest a few meeting times in the coming days, we can get something scheduled!

SteveMacenski commented:
Mid day in which time zone :wink: 

@acachathuranga what about you?

Daniel-Meza commented:
Good point 😅
Between 11am - 4pm CDT

acachathuranga commented:
Sorry for the late reply! Thanks for your time on this @SteveMacenski. I'm 12hrs ahead of your timezone haha, but actually I'm okay with any time this month. Let me know a convenient time for you, thanks!

SteveMacenski commented:
@Daniel-Meza could you do 10am CDT? That would be 8am my time, 10am your time, and 8pm @acachathuranga time -- which seems like the most reasonable cross section for everyone involved if you could do a little earlier. I'd be open Nov 20, 22, 26 (to name a few)

Daniel-Meza commented:
Yes, I can make 10am CDT on Nov 20 or 26.

SteveMacenski commented:
@Daniel-Meza you can find my email on my github page. Please email me so I have your email to set up the meeting. I see @acachathuranga has an email on his account that I can use

SteveMacenski commented:
Invites were sent out!

SteveMacenski commented:
Notes from our conversation:
- [x] Review this PR, make adjustments as needed
- [x] @acachathuranga add documentation on the initialization methods in detail
- [ ] @Daniel-Meza review it and test it 
- [ ] Steve: test it 
- [x] Obtain gif of robots / slam
- [ ] Merge baby merge :partying_face:

SteveMacenski commented on CMakeLists.txt:
```diff
@@ -64,6 +64,7 @@ set(libraries
     toolbox_common
     SlamToolboxPlugin
     ceres_solver_plugin
+    multirobot_slam_toolbox
```
For this + the Class + file names: I think we need to make this more specific since we'll have both nodes in the project. 

Suggestion: Decentralized Multi-robot?

acachathuranga replied:
Changed the class + file names with 'decentralized' prefix

SteveMacenski commented on README.md:
```diff
@@ -115,6 +115,37 @@ To minimize the amount of changes required for moving to this mode over AMCL, we
+
+*multirobot_slam_toolbox_node* extends slam_toolbox for multi-robot mapping and localization. Slam_toolbox instances are run on each robot under unique namespaces and pose graph data are shared between the nodes through */localized_scan* topic. Each slam_toolbox instance publishes transform of host robot and peers in the network. Peer transform frame names are prefixed with the namespace of the p ...
+
+Initially the robots should start close to each other, in the same orientation, as scan matching is used to find the connection between different pose graphs. This however is not required in simulation environments such as gazebo, as all robots' odometry is referenced to a gazebo's global frame. 
```
As discussed: expand on this. If you can show an example project setting up the shared frame, that would be immensely helpful for those that want to use your work

acachathuranga replied:
Expanded the documentation on decentralized slam. I moved the documentation to a separate file because I didn't want to clutter the main readme.

SteveMacenski commented on README.md:
```diff
@@ -115,6 +115,37 @@ To minimize the amount of changes required for moving to this mode over AMCL, we
+# Multi-Robot SLAM
+![multirobot_slam](/images/multi-robot_mapping.gif?raw=true "Multi-Robot SLAM")
+
+*multirobot_slam_toolbox_node* extends slam_toolbox for multi-robot mapping and localization. Slam_toolbox instances are run on each robot under unique namespaces and pose graph data are shared between the nodes through */localized_scan* topic. Each slam_toolbox instance publishes transform of host robot and peers in the network. Peer transform frame names are prefixed with the namespace of the p ...
```
I'd expand on this description about how the data is shared (localized scans published once processed by a robot's host node to sync up with others), why (network limitations). Enough so that when someone reads the software, they know what theyre looking at

SteveMacenski commented on include/slam_toolbox/laser_utils.hpp:
```diff
@@ -80,10 +80,12 @@ class LaserAssistant
     const std::string & base_frame);
   ~LaserAssistant();
   LaserMetadata toLaserMetadata(sensor_msgs::msg::LaserScan scan);
```
Should the single version be removed and uses switched to the new one?

SteveMacenski replied:
It looks like the single version just wraps the multi-version now

SteveMacenski commented on src/laser_utils.cpp:
```diff
@@ -144,34 +153,39 @@ karto::LaserRangeFinder * LaserAssistant::makeLaser(const double & mountingYaw)
 
 bool LaserAssistant::isInverted(double & mountingYaw)
 {
-  geometry_msgs::msg::TransformStamped laser_ident;
```
why was this method updated?

acachathuranga replied:
Previously this method was querying TF to to get laser to base_link transform. But in multi-robot scenario, TF is not broadcasted between robots (for bandwidth reasons). The transform is now shared through LocalizedLaserScan messages, so updated this method accordingly. I mentioned this in the updated documentation also.

SteveMacenski commented on src/slam_toolbox_multirobot.cpp:
```diff
@@ -0,0 +1,254 @@
+: SlamToolbox(options), localized_scan_topic_("/localized_scan")
+/*****************************************************************************/
+{
+    current_ns_ = this->get_namespace() + 1;
```
`get_namespace` returns a string, I'm not sure what this intends to do

SteveMacenski replied:
After reviewing the file, the use of `current_ns_` and its modification to the published localized range scan should be documented / explained further

acachathuranga replied:
> `get_namespace` returns a string, I'm not sure what this intends to do

get_namespace returns a char array, `+1` was to remove the leading `/` in namesapce. But now I changed this and the variable name as well, for clarity. Additional details about the namespace usage is added to the documentation

SteveMacenski commented on src/slam_toolbox_multirobot.cpp:
```diff
@@ -0,0 +1,254 @@
+
+/*****************************************************************************/
+MultiRobotSlamToolbox::MultiRobotSlamToolbox(rclcpp::NodeOptions options)
+: SlamToolbox(options), localized_scan_topic_("/localized_scan")
```
This should be a parameter (the topic to use)

acachathuranga replied:
done

SteveMacenski commented on src/slam_toolbox_multirobot.cpp:
```diff
@@ -0,0 +1,254 @@
+
+  sensor_msgs::msg::LaserScan::ConstSharedPtr scan = 
+    std::make_shared<sensor_msgs::msg::LaserScan>(localized_scan->scan);
+  Pose2 pose;
```
There's a `x,y,heading` constructor for Pose2 that would make this more concise https://github.com/SteveMacenski/slam_toolbox/blob/ros2/lib/karto_sdk/include/karto_sdk/Karto.h#L2063

acachathuranga replied:
done

SteveMacenski commented on src/slam_toolbox_multirobot.cpp:
```diff
@@ -0,0 +1,254 @@
+  }
+  LocalizedRangeScan * range_scan = addExternalScan(laser, scan, pose);
+  
+  if (range_scan != nullptr)
```
This is also publishing - aren't we having some confliction if this is also publishing?

https://github.com/SteveMacenski/slam_toolbox/blob/ros2/src/slam_toolbox_common.cpp#L472-L498

acachathuranga replied:
The slam_toolbox_common.cpp TF publisher is only publishing host robot transform. But here we're only publishing external robots' transforms.

SteveMacenski reviewed (commented):
It looks like this PR is in conflict with `ros2` now, that also needs to be resolved before I/Daniel can test. I also see a few linting issues, running `ament_cpplint` and `ament_uncrustify` will reveal them!

acachathuranga committed:
linting and style changes

acachathuranga commented:
Sorry, it had been a very busy couple weeks trying to wrap up work things before holidays, so couldn't finish all the changes. But the linting issues and the conflicts with ros2 branch are fixed now. 
Yet to be done,
- [ ] Update documentation for new multi-robot launch files in nav2-bringup
The new humble binaries for multi-robot launch in nav2-bringup seems to have some issue with parameters(Iron/Jazzy parameters back-ported humble I guess?). So, need to sort out how to guide others to easily run and play around with a multi-robot sim + slam. 
- [ ] Complete review changes

Sorry again for the delays as I'll be traveling starting tomorrow, so will take a while for me to finish the items. But will try my best to finish all before new year!

SteveMacenski commented:
No worries at all, its a busy time of year :-) Let us know when you're ready

MikeDegany commented:
I've been using this PR for collaborative mapping with my robots, and it's been working pretty well. Here's a video showing my robots in action using this code.
If there's anything I can do to help move this PR forward, I'd be happy to contribute. Please let me know!

https://github.com/user-attachments/assets/d3e887ee-f982-4d24-ac08-347dd6c15ff3

acachathuranga committed:
add decentralized specifier

acachathuranga committed:
decentralized multirobot class name update

acachathuranga committed:
convert decentralized multirobot node to lifecycle node

acachathuranga committed:
vector3 access fix for kilted

acachathuranga committed:
namespace variable refactor for clarity

acachathuranga committed:
get localizedscan topic from parameter

acachathuranga committed:
cleanup Pose2 creation

acachathuranga committed:
odom frame update to match kilted tb3 sim

acachathuranga committed:
readme update for decentralized multirobot slam

acachathuranga committed:
comment on laser util updates

acachathuranga committed:
add back missing dependencies

SteveMacenski commented on docs/Screenshot from 2025-10-05 06-11-02.png:
Isn't this the same as the node graph below?

acachathuranga replied:
my bad, accidentally included an extra image

SteveMacenski commented on README.md:
```diff
@@ -111,6 +111,19 @@ To minimize the amount of changes required for moving to this mode over AMCL, we
 
 In summary, this approach I dub `elastic pose-graph localization` is where we take existing map pose-graphs and localized with-in them with a rolling window of recent scans. This way we can localize in an existing map using the scan matcher, but not update the underlaying map long-term should something go wrong. It can be considered a replacement to AMCL and results is not needing any .pgm maps e ...
 
+# Multi-Robot SLAM
```
Put above lifelong mapping since this is less experimental than those items

acachathuranga replied:
moved multirobot readme section above lifelong mapping

SteveMacenski commented on src/laser_utils.cpp:
```diff
@@ -85,6 +85,16 @@ LaserMetadata LaserAssistant::toLaserMetadata(sensor_msgs::msg::LaserScan scan)
 {
   scan_ = scan;
   frame_ = scan_.header.frame_id;
+  return toLaserMetadata(scan, readLaserPose());
```
Can we make this a default argument value so we have only 1 implementation of `toLaserMetadata`? Or, otherwise can we consolidate this into one implementation?

acachathuranga replied:
consolidated both implementations using a default argument. I doubt we'll be backporting this to an earlier version than C++17? hence used std::optional for the default argument.

SteveMacenski commented on src/laser_utils.cpp:
```diff
@@ -95,9 +105,10 @@ LaserMetadata LaserAssistant::toLaserMetadata(sensor_msgs::msg::LaserScan scan)
   karto::LaserRangeFinder * laser =
     karto::LaserRangeFinder::CreateLaserRangeFinder(
-    karto::LaserRangeFinder_Custom, karto::Name("Custom Described Lidar"));
+    karto::LaserRangeFinder_Custom, karto::Name("Custom Described Lidar : " + laser_namespace));
```
If there is no forward slash, then this will return the full name, no?

acachathuranga replied:
ya the idea was that the host robot will use the `frame id` as `laser name`. In contrast, external lasers will use `namespace` as `laser name`. But I realized this could be confusing, therefore updated it. Now all lasers will use `full frame id` as laser name. Eg:
> Registering sensor: [Custom Described Lidar : base_scan]  &emsp;  <- Register host scanner
> Registering sensor: [Custom Described Lidar : robot2 base_scan] &emsp; <- Register another robot's scanner

acachathuranga committed:
removed extra image

acachathuranga committed:
readme update for multirobot

acachathuranga committed:
consolidate multiple laser metadata implementations

acachathuranga committed:
use laser full frame_id as laser name

SteveMacenski commented on src/slam_toolbox_decentralized_multirobot.cpp:
```diff
@@ -0,0 +1,283 @@
+    return;
+  }
+
+  LocalizedRangeScan * range_scan = addScan(laser, scan, pose);
```
shouldProcessScan? https://github.com/SteveMacenski/slam_toolbox/blob/a1ecfea8f242206dc105fc930299bc8cb63783e9/src/slam_toolbox_async.cpp#L57

acachathuranga replied:
Added `shouldProcessScan` check before processing host scans. Now host slam can be paused. But for multi-robot slam, the external scans will still be processed. Can't say for sure that this is the ideal behavior for all situations... but I thought it's better to just condition host scan processing.

acachathuranga committed:
shouldProcessScan check for host scans

acachathuranga committed:
minor readme corrections

SteveMacenski reviewed (approved):
Anything else you think should be updated before a merge?

SteveMacenski commented:
@Daniel-Meza are you interested in reviewing it and testing it

acachathuranga commented:
> Anything else you think should be updated before a merge?

Thanks for the reviews, Steve! Everything looks complete from my end. Once you or Daniel have tested and confirmed things look good, we can proceed with the merge

SteveMacenski commented:
Given ROSCon is around the corner and I want you to be able to say that this is in the project, I'm going to merge this in! I may ping you if I receive tickets / PRs for this part of the codebase as the principle author, but I don't expect much traffic until after you announce this at ROSCon :wink: 

Thanks for the follow up and I look forward to speaking with you live in Singapore next week. Lets get a beer.

\- Steve

acachathuranga commented:
Thanks a lot, Steve, really appreciate your feedback and the detailed reviews. Excited to see it merged and part of the project! 
I'll be ready to support if any follow-ups come in. 
Looking forward to catching up in Singapore next week! - Achala

Amronos commented:
Hello @acachathuranga @SteveMacenski, is a backport to kilted or jazzy possible? Thanks for this great feature!

SteveMacenski commented:
This will not be backported as it breaks API/ABI stability in a couple of the functions used for folks' work. It will be available in Lyrical and newer though in binaries -- otherwise you should be able to compile this branch on Jazzy/Kilted to get it from source :-)
